api/model/src/main/java/org/projectnessie/model/Elements.java

```diff
@@ -0,0 +1,203 @@
+   *
+   * <p>The returned value is always processable by Nessie services announcing Nessie spec 2.2.0 or
+   * newer.
+   *
+   * @return The URI path compatible representation of the given elements, possibly escaped. The
+   *     returned value should be URL-encoded before added to a URI path. The returned value can be
+   *     parsed with {@link #elementsFromPathString(String)}.
+   */
+  default String toPathStringEscaped() {
+    return Util.toPathStringEscaped(getElements());
+  }
+
+  /**
+   * Escapes content-key elements into a canonical form that escapes {@code .} characters. The
+   * returned format is similar to {@linkplain #toPathStringEscaped()}, but does not escape
+   * problematic URI characters.
+   *
+   * <p>Algorithm:
+   *
+   * <ul>
+   *   <li>Elements are separated using a single {@code .} character when <em>not</em> escaping,
+   *   <li>Elements are separated using a two dots {@code ...} characters when escaping,
+   *   <li>If an element starts with a {@code .} character, that element <b>and all following
+   *       elements</b> use the "dot character escaping":
+   *       <ul>
+   *         <li>a {@code .} is escaped as {@code ._}
+   *       </ul>
+   * </ul>
+   *
+   * <p>Some examples:
+   *
+   * <ul>
+   *   <li>{@code ["foo", "bar", "baz"]} returned as {@code "foo.bar.baz"} - no escaping needed.
+   *   <li>{@code ["foo", ".bar", "baz"]} returned as {@code "foo..._bar..baz"} - escaping needed
+   *       with the 2nd element. The first dot is the element separator. The 2nd dot is the first
+   *       character of the second element, indicating that escaping starts at this element. {@code
+   *       ._} is the escape sequence for the {@code .} character.
+   *   <li>{@code ["foo.", ".bar", "a/\%aa"]} returned as {@code ".foo._.._bar..a/\%aa"}
+   *   <li>{@code ["foo", "bar", "a/\%aa"]} returned as {@code "foo.bar...a/\%aa"}
+   * </ul>
+   *
+   * @return The canonical representation of the given elements, possibly escaped. The returned
+   *     value should <em>not</em> be used in a URI path. The returned value can be parsed with
+   *     {@link #elementsFromPathString(String)}.
+   */
+  default String toCanonicalString() {
+    return Util.toCanonicalString(getElements());
+  }
+
+  default void validate(String type) {
+    List<String> elements = getElements();
+    if (elements.size() > MAX_ELEMENTS) {
+      throw new IllegalStateException(
+          format("%s too long, max allowed number of elements: %s", type, MAX_ELEMENTS));
+    }
+    int sum = 0;
+    for (String e : elements) {
+      if (e == null) {
+        throw new IllegalArgumentException(
+            format("%s '%s' must not contain a null element", type, elements));
+      }
+      int l = e.length();
+      if (l == 0) {
+        throw new IllegalArgumentException(
+            format("%s '%s' must not contain an empty element", type, elements));
+      }
+      if (".".equals(e)) {
+        throw new IllegalArgumentException(
```

Comment:
This was previously thrown only for the last element of a `Namespace`. It wasn't thrown for `ContentKey` at all, and also not for non-last elements of a `Namespace`.

Reply:
Maybe you could remove this check here and add it in `Namespace.validate()`, then you don't need the `boolean nsSingleDot` parameter, wdyt?

```java
  @Override
  public void validate(String type) {
    Elements.super.validate(type);
    String last = getElements().get(getElements().size() - 1);
    if (DOT_STRING.equals(last)) {
      throw new IllegalArgumentException(
          format("Namespace '%s' must not contain a '.' element", toCanonicalString()));
    }
  }
```